elfi/methods/utils.py

```diff
@@ -184,29 +184,70 @@ def logpdf(cls, x, means, cov=1, weights=None):
         return np.log(cls.pdf(x, means=means, cov=cov, weights=weights))
 
     @classmethod
-    def rvs(cls, means, cov=1, weights=None, size=1, random_state=None):
+    def rvs(cls, means, cov=1, weights=None, size=1, prior_logpdf=None, random_state=None):
         """Draw random variates from the distribution.
 
         Parameters
         ----------
         means : array_like
             means of the Gaussian mixture components
-        weights : array_like
-            1d array of weights of the gaussian mixture components
-        cov : array_like
+        cov : array_like, optional
             a shared covariance matrix for the mixture components
-        size : int or tuple
-        random_state : np.random.RandomState or None
+        weights : array_like, optional
+            1d array of weights of the gaussian mixture components
+        size : int or tuple, optional
+        prior_logpdf : callable, optional
+            Can be used to check validity of random variable.
+        random_state : np.random.RandomState, optional
 
         """
-        means, weights = cls._normalize_params(means, weights)
         random_state = random_state or np.random
+        means, weights = cls._normalize_params(means, weights)
 
-        inds = random_state.choice(len(means), size=size, p=weights)
-        rvs = means[inds]
-        perturb = ss.multivariate_normal.rvs(
-            mean=means[0] * 0, cov=cov, random_state=random_state, size=size)
-        return rvs + perturb
+        # additional complexity due to supporting scipy-like ambiguity of arguments
+        if size is None:
+            size = means.shape[0]
+            output = []
+            scalar = True
+        else:
+            output = np.empty((size,) + means.shape[1:])
+            scalar = False
+
+        n_accepted = 0
+        n_left = size
+        trials = 0
+
+        while n_accepted < size:
+            inds = random_state.choice(len(means), size=n_left, p=weights)
+            rvs = means[inds]
+            perturb = ss.multivariate_normal.rvs(mean=means[0] * 0,
+                                                 cov=cov,
+                                                 random_state=random_state,
+                                                 size=n_left)
+            x = rvs + perturb
+
+            # check validity of x
+            if prior_logpdf is not None:
+                x = x[np.isfinite(prior_logpdf(x))]
+                n_accepted1 = len(x)
+            else:
```

Comment:
This can be removed. It should be always valid to say `n_accepted1 = len(x)`. The name `n_accepted1` is a bit unclear.

Reply:
`x` could be a scalar, and casting to array seems unnecessary.

Hmm... I find `n_accepted1` much clearer for this purpose than `n_new`... :D

Reply:
It won't be a scalar since size is an integer.

Reply:
Oh yes, that's true now.